### ISSUE TITLE
[BUG] One member of Multi-node Etcd start its own cluster as its data-dir structure is invalid which triggers the restoration of single member.

### ISSUE THREAD

**Issue by timuthy** [gardener/etcd-druid#361]
**Describe the bug:**
When creating a new `etcd` multi-node cluster it can happen that **multiple clusters** are created instead of one cluster with multiple members. As a result, consumers get a single `etcd-client` service which is backed by distinct clusters.

Etcd-Druid and all involved components use a static bootstrapping procedure as described [here](https://etcd.io/docs/v3.4/op-guide/clustering/#static).
In combination with the StatefulSet's [parallel podManagementPolicy](https://github.com/gardener/etcd-druid/blob/4da816d7247971c5f47cb529b7df2894cb21e1bd/charts/etcd/templates/etcd-statefulset.yaml#L19) it sometimes happen that two or more pods don't detect an existing cluster which makes them to create a new cluster (due to `--initial-cluster-state new`) at the same time.

In this example we saw it for two pods which belonged to the same StatefulSet (please also note the version difference):

- etcd-main-0
```
2022-06-24T15:08:15.87671756Z stderr F 2022-06-24 15:08:15.876585 I | etcdserver: setting up the initial cluster version to 3.0
```

- etcd-main-2
```
2022-06-24T15:08:23.391710822Z stderr F 2022-06-24 15:08:23.391621 I | etcdserver: setting up the initial cluster version to 3.4
```

After all the pods were running we constantly observed logs like:

```
2022-06-24T15:08:25.488303476Z stderr F 2022-06-24 15:08:25.488156 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[1e2acb0b00e637d5]=cdf818194e3a8c32, local=dd673093ade40ad6)
2022-06-24T15:08:25.488463226Z stderr F 2022-06-24 15:08:25.488363 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[1e2acb0b00e637d5]=cdf818194e3a8c32, local=dd673093ade40ad6)
2022-06-24T15:08:25.565269748Z stderr F 2022-06-24 15:08:25.564967 E | rafthttp: request sent was ignored (cluster ID mismatch: remote[1e2acb0b00e637d5]=cdf818194e3a8c32, local=dd673093ade40ad6)
2022-06-24T15:08:25.586832244Z stderr F 2022-06-24 15:08:25.586699 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[1e2acb0b00e637d5]=cdf818194e3a8c32, local=dd673093ade40ad6)
```

**Expected behavior:**
Etcd-Druid to create a single cluster with the amount of replicas configured in the `etcd` resource.

**How To Reproduce (as minimally and precisely as possible):**
Create a mutli-node `etcd` cluster with Etcd-Druid.

**Environment (please complete the following information):**
 - Etcd-druid version/commit ID : `v0.10.0`

**Anything else we need to know?**:
We analyzed this issue together with @rfranzke and @timebertt and in addition checked how [KinD](https://github.com/kubernetes-sigs/kind/) bootstraps multi-node etcd clusters. Turned out that they create a new cluster once with a single member and then scale-up by one member at a time:

- Pod-0
```
spec:
  containers:
  - command:
    - etcd
    - --advertise-client-urls=https://172.18.0.4:2379
    - --cert-file=/etc/kubernetes/pki/etcd/server.crt
    - --client-cert-auth=true
    - --data-dir=/var/lib/etcd
    - --experimental-initial-corrupt-check=true
    - --initial-advertise-peer-urls=https://172.18.0.4:2380
    - --initial-cluster=test-control-plane=https://172.18.0.4:2380
    - --key-file=/etc/kubernetes/pki/etcd/server.key
    - --listen-client-urls=https://127.0.0.1:2379,https://172.18.0.4:2379
    - --listen-metrics-urls=http://127.0.0.1:2381
    - --listen-peer-urls=https://172.18.0.4:2380
    - --name=test-control-plane
    - --peer-cert-file=/etc/kubernetes/pki/etcd/peer.crt
    - --peer-client-cert-auth=true
    - --peer-key-file=/etc/kubernetes/pki/etcd/peer.key
    - --peer-trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
    - --snapshot-count=10000
    - --trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
```

- Pod-1
```
spec:
  containers:
  - command:
    - etcd
    - --advertise-client-urls=https://172.18.0.2:2379
    - --cert-file=/etc/kubernetes/pki/etcd/server.crt
    - --client-cert-auth=true
    - --data-dir=/var/lib/etcd
    - --experimental-initial-corrupt-check=true
    - --initial-advertise-peer-urls=https://172.18.0.2:2380
    - --initial-cluster=test-control-plane=https://172.18.0.4:2380,test-control-plane2=https://172.18.0.2:2380
    - --initial-cluster-state=existing
    - --key-file=/etc/kubernetes/pki/etcd/server.key
    - --listen-client-urls=https://127.0.0.1:2379,https://172.18.0.2:2379
    - --listen-metrics-urls=http://127.0.0.1:2381
    - --listen-peer-urls=https://172.18.0.2:2380
    - --name=test-control-plane2
    - --peer-cert-file=/etc/kubernetes/pki/etcd/peer.crt
    - --peer-client-cert-auth=true
    - --peer-key-file=/etc/kubernetes/pki/etcd/peer.key
    - --peer-trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
    - --snapshot-count=10000
    - --trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
```
- Pod-2
```
spec:
  containers:
  - command:
    - etcd
    - --advertise-client-urls=https://172.18.0.3:2379
    - --cert-file=/etc/kubernetes/pki/etcd/server.crt
    - --client-cert-auth=true
    - --data-dir=/var/lib/etcd
    - --experimental-initial-corrupt-check=true
    - --initial-advertise-peer-urls=https://172.18.0.3:2380
    - --initial-cluster=test-control-plane3=https://172.18.0.3:2380,test-control-plane=https://172.18.0.4:2380,test-control-plane2=https://172.18.0.2:2380
    - --initial-cluster-state=existing
    - --key-file=/etc/kubernetes/pki/etcd/server.key
    - --listen-client-urls=https://127.0.0.1:2379,https://172.18.0.3:2379
    - --listen-metrics-urls=http://127.0.0.1:2381
    - --listen-peer-urls=https://172.18.0.3:2380
    - --name=test-control-plane3
    - --peer-cert-file=/etc/kubernetes/pki/etcd/peer.crt
    - --peer-client-cert-auth=true
    - --peer-key-file=/etc/kubernetes/pki/etcd/peer.key
    - --peer-trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
    - --snapshot-count=10000
    - --trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
```

Thanks for you input @timebertt and @rfranzke.

**Comment by ishan16696** [comment]
TBH, I'm pretty confused with the given information/evidences. These are my initial thoughts:
Isn’t Leader Election have to racy as it works on **greedy algorithm** i.e one cluster member triggers an election, voted for himself and wants other members in cluster to vote for him as well ?

Now, the question comes up whether other members of cluster are reachable/available or not. 
If they are not reachable/available then it will not become leader as it will never be able to attain majority of votes.

> it sometimes happen that two or more pods don't detect an existing cluster which makes them to create a new cluster (due to --initial-cluster-state new)

How it can start its own cluster. According to RAFT, it should triggered a leader-election and keep waiting to get votes provided they got the correct configuration. It could be the configuration issue which leads them to start a new cluster.
Have you tried to enter to a `etcd` container and checked the member list to understand what config etcd member got ?


> ```2022-06-24T15:08:15.87671756Z stderr F 2022-06-24 15:08:15.876585 I | etcdserver: setting up the initial cluster version to 3.0```
```2022-06-24T15:08:23.391710822Z stderr F 2022-06-24 15:08:23.391621 I | etcdserver: setting up the initial cluster version to 3.4```

Hmm, these logs are suspicious. I tried running `multi-node etcdbackup-restore` in a kind cluster. It works fine but I saw logs like this:
```
2022-06-28 04:56:42.748140 I | etcdserver: starting server... [version: 3.4.13, cluster version: to_be_decided]
.
.
2022-06-28 04:56:42.981296 N | etcdserver/membership: set the initial cluster version to 3.0
2022-06-28 04:56:42.981368 I | etcdserver/api: enabled capabilities for version 3.0
.
.
2022-06-28 04:57:15.025421 N | etcdserver/membership: updated the cluster version from 3.0 to 3.4
```

whereas I also tried to run multi-node etcd-backup-restore (not on kinD).
```
k logs etcd-main-2 -c etcd | grep "version"                       
2022-06-28 05:27:13.047746 I | etcdserver: starting server... [version: 3.4.13, cluster version: to_be_decided]
.
2022-06-28 05:27:13.071696 N | etcdserver/membership: set the initial cluster version to 3.4
.
2022-06-28 05:27:13.071763 I | etcdserver/api: enabled capabilities for version 3.4
```
As you I see it don't have any such logs related to `cluster version to 3.0` when I ran multi-node etcd-backup-restore not on KinD.

**Comment by ishan16696** [comment]
Just to mention @unmarshall , @aaronfern and I saw one issue in which etcd cluster members were unable to reach out to each other (may be due some DNS issue) for more than hour.
My point is even though `etcd cluster members were unable to reach out to each other >1 hour` but no etcd member in a cluster had started its own cluster.

**Comment by ishan16696** [comment]
@abdasgupta @aaronfern  Are we using `--initial-cluster-token` anywhere in config map which provides configuration to etcd ?

**Comment by aaronfern** [comment]
@ishan16696 yes, we do use `--initial-cluster-token`
It's added to the configmap by etcd-druid [here](https://github.com/gardener/etcd-druid/blob/4da816d7247971c5f47cb529b7df2894cb21e1bd/pkg/component/etcd/configmap/configmap.go#L162)

**Comment by ishan16696** [comment]
> It's added to the configmap by etcd-druid [here](https://github.com/gardener/etcd-druid/blob/4da816d7247971c5f47cb529b7df2894cb21e1bd/pkg/component/etcd/configmap/configmap.go#L162)

ok, but seems like it is hardcoded. I will update to use some unique identifier.

**Comment by timuthy** [comment]
> --initial-cluster-token can act a unique identifier for each cluster. Specifying this can protect us from unintended cross-cluster interaction. But I'm not very sure that using that flag will be helpful in this case or not.

Can you explain how that setting helps to mitigate the issue we saw? The documentation says:

> If you are spinning up multiple clusters (or creating and destroying a single cluster) with same configuration for testing purpose, it is highly recommended that you specify a unique initial-cluster-token for the different clusters. By doing this, etcd can generate unique cluster IDs and member IDs for the clusters even if they otherwise have the exact same configuration. This can protect you from cross-cluster-interaction, which might corrupt your clusters.

And

> "Initial cluster token for the etcd cluster during bootstrap. Specifying this can protect you from unintended cross-cluster interaction when running multiple clusters."

In Gardener's setup each cluster has a unique set of members which should prevent cross cluster communication from the start.

> My point is even though etcd cluster members were unable to reach out to each other >1 hour but no etcd member in a cluster had started its own cluster(they also have --initial-cluster-state new)

Makes sense, thanks for explaining 👍 
Then we have to dig further and search for another root cause? Changing `--initial-cluster-state` doesn't sound promising to me at the moment.

**Comment by ishan16696** [comment]
Thanks @timuthy for providing the logs of `e2e test of HA shoots` which is backed by multi-node `etcd-backup-restore` cluster.
I take a look at backup-restore logs of each cluster member. 
In **brief**, it is happening due to PVC of 1 etcd cluster member got corrupted.

**Details:**
Initially all 3 cluster members have started normally.

- backup-restore-1 becomes the backup-Leader:
  ```
  2022-06-24T15:08:15.87734049Z stderr F time="2022-06-24T15:08:15Z" level=info msg="backup-restore became: Leader" actor=leader-elector
  2022-06-24T15:08:15.87737824Z stderr F time="2022-06-24T15:08:15Z" level=info msg="backup-restore started leading..." actor=leader-elector
  2022-06-24T15:08:15.877384669Z stderr F time="2022-06-24T15:08:15Z" level=info msg="Creating snapstore from provider: Local" actor=backup-restore-server
  ```
- backup-restore-2 becomes the backup-Follower:
  ```
  2022-06-24T15:08:18.556259624Z stderr F time="2022-06-24T15:08:18Z" level=info msg="backup-restore changed the state from UnknownState to Follower" actor=leader-elector
  ```

- backup-restore-0:  
   - it started normally with `initialization` request
   ```
  2022-06-24T15:07:51.424618726Z stderr F time="2022-06-24T15:07:51Z" level=info msg="No snapshot found. Will remove the data directory."
  2022-06-24T15:07:51.424624406Z stderr F time="2022-06-24T15:07:51Z" level=info msg="Removing directory(/var/etcd/data/new.etcd) since snapstore is empty."
  2022-06-24T15:07:51.424626676Z stderr F time="2022-06-24T15:07:51Z" level=info msg="Successfully initialized data directory for etcd." actor=backup-restore-server
  2022-06-24T15:07:51.437644328Z stderr F time="2022-06-24T15:07:51Z" level=info msg="Responding to status request with: Successful" 
  ```
  - But then I think etcd-0's etcd container got restarted and backup-restore-0 got `Initialization` request. 
  ```
  2022-06-24T15:08:20.550396624Z stderr F time="2022-06-24T15:08:20Z" level=info msg="backup-restore is in: UnknownState" actor=leader-elector
  2022-06-24T15:08:20.550401914Z stderr F time="2022-06-24T15:08:20Z" level=info msg="waiting for Re-election..." actor=leader-elector
  2022-06-24T15:08:21.76978254Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Responding to status request with: New" actor=backup-restore-server
  2022-06-24T15:08:21.78623113Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Received start initialization request." actor=backup-restore-server
  ```
  - But till that time `backup-restore-1` (backup-leader) has already uploaded the 1 full snapshot.
  ```
  2022-06-24T15:08:16.938618906Z stderr F time="2022-06-24T15:08:16Z" level=info msg="Taking scheduled full snapshot for time: 2022-06-24 15:08:16.938451926 +0000 UTC" actor=snapshotter
  .
  2022-06-24T15:08:16.949699989Z stderr F time="2022-06-24T15:08:16Z" level=info msg="Total time taken by Snapshot API: 0.004520 seconds." actor=snapshotter
  2022-06-24T15:08:16.949707659Z stderr F time="2022-06-24T15:08:16Z" level=info msg="start compressing the snapshot using gzip Compression Policy" actor=compressor
  .
  2022-06-24T15:08:16.953126307Z stderr F time="2022-06-24T15:08:16Z" level=info msg="Successfully saved full snapshot at: Full-00000000-00000001-1656083296.gz" actor=snapshotter
  ```
   - While Initialization of etcd, backup-retore-0 [found the data-dir structure](https://github.com/gardener/etcd-backup-restore/blob/ab51492d39a98900e038506fc8dfb97ead3c0acb/pkg/initializer/validator/datavalidator.go#L130) is **Invalid**.
  ```
    2022-06-24T15:08:21.78639462Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Checking for data directory structure validity..."
    2022-06-24T15:08:21.786406809Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Data directory structure invalid."
  ```
   - Hence, It triggers the restoration and backup-restore-0 restores from the full snapshot which was uploaded by backup-leader(i.e backup-restore-1).
  ```
      2022-06-24T15:08:21.7864388Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Finding latest set of snapshot to recover from..."
      2022-06-24T15:08:21.786545639Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Removing directory(/var/etcd/data/new.etcd.part)."
      2022-06-24T15:08:21.78660167Z stderr F time="2022-06-24T15:08:21Z" level=info msg="Restoring from base snapshot: Full-00000000-00000001-1656083296.gz" actor=restorer
      2022-06-24T15:08:21.786889539Z stderr F time="2022-06-24T15:08:21Z" level=info msg="start decompressing the snapshot with gzip compressionPolicy" actor=de-compressor
      2022-06-24T15:08:21.791924926Z stderr F time="2022-06-24T15:08:21Z" level=info msg="successfully fetched data of base snapshot in 0.005019517 seconds [CompressionPolicy:gzip]" actor=restorer
    ```
  - Hence metadata of corresponding etcd of backup-restore-0 got wrong hence it is unable to join the cluster and forms its own cluster.

**Comment by ishan16696** [comment]
The support for `single member restoration in multi-node etcd` has been added and released with [backup-restore v0.19.0](https://github.com/gardener/etcd-backup-restore/releases/tag/v0.19.0)
This issue now has been permanently closed by adding the feature of single member restoration in multi-node etcd, please refer this PR: https://github.com/gardener/etcd-backup-restore/pull/509
